paparazzi/src/main/java/app/cash/paparazzi/PaparazziTestExtension.kt

```diff
@@ -0,0 +1,465 @@
+import com.android.ide.common.rendering.api.RenderSession
+import com.android.ide.common.rendering.api.Result
+import com.android.ide.common.rendering.api.Result.Status.ERROR_UNKNOWN
+import com.android.ide.common.rendering.api.SessionParams
+import com.android.layoutlib.bridge.Bridge.cleanupThread
+import com.android.layoutlib.bridge.Bridge.prepareThread
+import com.android.layoutlib.bridge.BridgeRenderSession
+import com.android.layoutlib.bridge.impl.RenderAction
+import com.android.layoutlib.bridge.impl.RenderSessionImpl
+import com.android.tools.layoutlib.java.System_Delegate
+import org.junit.jupiter.api.extension.AfterEachCallback
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+import java.awt.image.BufferedImage
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier
+import java.util.*
+import java.util.concurrent.TimeUnit
+import kotlin.collections.LinkedHashSet
+import kotlin.collections.find
+import kotlin.collections.forEach
+import kotlin.collections.setOf
+
+class PaparazziTestExtension(
+  private val environment: Environment = detectEnvironment(),
+  private val deviceConfig: DeviceConfig = DeviceConfig.NEXUS_5,
+  private val theme: String = "android:Theme.Material.NoActionBar.Fullscreen",
+  private val appCompatEnabled: Boolean = true,
+  private val maxPercentDifference: Double = 0.1,
+  private val snapshotHandler: SnapshotHandler = determineHandler(maxPercentDifference)
+) : BeforeEachCallback, AfterEachCallback {
+  private val THUMBNAIL_SIZE = 1000
+
+  private val logger = PaparazziLogger()
+  private lateinit var sessionParamsBuilder: SessionParamsBuilder
+  private lateinit var renderer: Renderer
+  private lateinit var renderSession: RenderSessionImpl
+  private lateinit var bridgeRenderSession: RenderSession
+  private var testName: TestName? = null
+  private val renderExtensions = LinkedHashSet<RenderExtension>()
+
+  val layoutInflater: LayoutInflater
+    get() = RenderAction.getCurrentContext().getSystemService("layout_inflater") as BridgeInflater
+
+  val resources: Resources
+    get() = RenderAction.getCurrentContext().resources
+
+  val context: Context
+    get() = RenderAction.getCurrentContext()
+
+  val contentRoot = """
+        |<?xml version="1.0" encoding="utf-8"?>
+        |<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+        |              android:layout_width="match_parent"
+        |              android:layout_height="match_parent"/>
+        """.trimMargin()
+
+  override fun beforeEach(context: ExtensionContext) {
+    registerFontLookupInterceptionIfResourceCompatDetected()
+    registerViewEditModeInterception()
+    registerMatrixMultiplyInterception()
+
+    InterceptorRegistrar.registerMethodInterceptors()
+
+    prepare(context)
+  }
+
+  override fun afterEach(context: ExtensionContext) {
+    close()
+
+    InterceptorRegistrar.clearMethodInterceptors()
+  }
```

Comment:
Where JUnit 4 `TestRule`s wrapped the test functions, JUnit 5 test extensions leverage callbacks to do the same work before & after instead.